### ISSUE TITLE
Background threads not exiting after calling loss.backward() on Win7

### ISSUE THREAD

**Issue by frebro83** [pytorch/pytorch#48888]
## 🐛 Bug

Background threads not exiting after calling loss.backward() on Win7, forced to kill the python console between each run.

## To Reproduce

1. Run conda with python 3.8 on windows 7
2. Run on either cpu or gpu (both tried)
3. Install from wheel, pip or build from source (all three tried)
3. Calculate a loss based on your loss function
4. Call loss.backward()
5. Program will continue until the end of the script but never give back the command prompt

```
import torch
import torch.nn as nn
import numpy as np

dtype = torch.float
device = torch.device("cpu")

class MLP(nn.Module):
    def __init__(self, inshape, outshape):
        super(MLP, self).__init__()
        self.layers = nn.Sequential(
            nn.Linear(inshape, outshape)
        )

    def forward(self, x):
        x = x.view(x.size(0), -1)
        x = self.layers(x)
        return x

data = torch.from_numpy(np.random.rand(1, 10).astype(np.float32))
target = torch.from_numpy(np.random.rand(1, 1).astype(np.float32))

model = MLP(10, 1)
optimizer = torch.optim.Adam(model.parameters(), lr=0.001)
loss_fn = nn.MSELoss()

predict = model(data)
loss = loss_fn(predict, target)

model.zero_grad()
loss.backward()
optimizer.step()

print("Program done, command prompt expected below.")
```
returns
```
(PY38) D:\GitData>C:/Users/Grinay/Anaconda3/envs/PY38/python.exe d:/GitData/bitnet/train_net.py
Program done, command prompt expected below.

```
and nothing happens until I kill the python terminal.

## Expected behavior

Expected behaviour is that the code finishes to run until done and then returns to the command prompt.

## Environment

```
(PY38) D:\GitData\pytorch\torch\utils>python collect_env.py
Collecting environment information...
PyTorch version: 1.8.0a0+6317e0b
Is debug build: False
CUDA used to build PyTorch: 10.2
ROCM used to build PyTorch: N/A

OS: Microsoft Windows 7 Professional N
GCC version: Could not collect
Clang version: Could not collect
CMake version: version 3.18.2

Python version: 3.8 (64-bit runtime)
Is CUDA available: True
CUDA runtime version: Could not collect
GPU models and configuration: GPU 0: GeForce GTX 780
Nvidia driver version: 441.22
cuDNN version: C:\Program Files\NVIDIA GPU Computing Toolkit\CUDA\v10.2\bin\cudnn64_7.dll
HIP runtime version: N/A
MIOpen runtime version: N/A

Versions of relevant libraries:
[pip3] numpy==1.19.4
[pip3] torch==1.8.0a0
[pip3] torchaudio==0.7.0
[pip3] torchvision==0.8.1
[conda] blas                      1.0                         mkl
[conda] cudatoolkit               10.2.89              h74a9793_1
[conda] mkl                       2019.0                   pypi_0    pypi
[conda] mkl-include               2020.2                      256
[conda] mkl-service               2.3.0            py38h2bbff1b_0
[conda] mkl_fft                   1.2.0            py38h45dec08_0
[conda] mkl_random                1.1.1            py38h47e9c7a_0
[conda] numpy                     1.19.4                   pypi_0    pypi
[conda] numpy-base                1.19.2           py38ha3acd2a_0
[conda] torch                     1.8.0a0                  pypi_0    pypi
[conda] torchaudio                0.7.0                      py38    pytorch
[conda] torchvision               0.8.1                py38_cu102    pytorch

(PY38) D:\GitData\pytorch\torch\utils>
```

 - PyTorch Version (e.g., 1.0): Multiple versions tested
 - OS (e.g., Linux): Win7 Pro N
 - How you installed PyTorch (`conda`, `pip`, source): conda, pip wheel and source tested
 - Build command you used (if compiling from source): Ninja with VS2019, cuda, cudnn (Edit: and compute_arch 3.5 for my GPU)
 - Python version: 3.8 (I think I tested with 3.7 as well, I do not remember though)
 - CUDA/cuDNN version: 10.2/7.6.XXX, (Edit: also tested from wheel and conda without cuda/cudnn)
 - GPU models and configuration: GTX780 with the drivers shipped with CUDA 10.2 (Edit: also tested with the newest graphic drivers)
 - Any other relevant information: Sorry, none.

## Additional context

Edit: Commenting out the loss.backward() call in the code-snippet will make the program exit normally.

cc @peterjc123 @maxluk @nbcsm @guyang3532 @gunandrose4u @mszhanyi @skyline75489

**Comment by peterjc123** [comment]
Related: https://github.com/pytorch/pytorch/issues/29383
Sadly, there is no workground available now.

**Comment by peterjc123** [comment]
I found a related article about it. https://stackoverflow.com/questions/60452341/stdconditional-variable-notify-problem-on-windows-7-called-from-destructor-o Looks like we have two options to fix this:
1. Switch to another implementation of `mutex` or `conditional_variable`.
2. Refactor the code to avoid calling `notify` in destructors.
What do you think, @ezyang @malfet ?

**Comment by ezyang** [comment]
Let's just leak the threads on shutdown. It's been causing other problems on non-Windows too. It would basically involve reverting https://github.com/pytorch/pytorch/pull/21438 Does someone want to give it a try?

**Comment by gccollect** [comment]
Any chance of this change getting merged into master?

**Comment by ezyang** [comment]
IIRC @malfet landed an alternate fix that is supposed to have subsumed this. Is this still failing on the latest release of PyTorch?

**Comment by gccollect** [comment]
It is still not existing using the latest nightly on windows 7

**Comment by gccollect** [comment]
Thanks @ezyang, adding the code below fixes the issue:
`os.environment["TORCH_AUTOGRAD_SHUTDOWN_WAIT_LIMIT"] = "0"`

**Comment by qaz2883383** [comment]
> Thanks @ezyang, adding the code below fixes the issue:
> `os.environment["TORCH_AUTOGRAD_SHUTDOWN_WAIT_LIMIT"] = "0"`

Thanks a lot. My program kill its own process well after adding code below.

`import os`
`os.environ["TORCH_AUTOGRAD_SHUTDOWN_WAIT_LIMIT"] = "0"`

**Comment by Sopel97** [comment]
Another workaround that works for all versions, though produces ugly output

```
if sys.platform == "win32":
    os.system(f'wmic process where processid="{os.getpid()}" call terminate >nul')
```

**Comment by Blackhex** [comment]
@malfet This seems to be resolved, is there anything more to do, or we can close it?

**Comment by vladimir-aubrecht** [comment]
Closing as issue seems to be stale and fixed. Please create new issue in case problem persists.
